### ISSUE TITLE
False positive on "reportOptionalMemberAccess"

### ISSUE THREAD

**Issue by belliriccardo** [microsoft/pyright#8959]
Consider the following example code (minimal reproducible example):

```python
from typing import Optional
import random

class Dummy:
    def __init__(self, value: int) -> None:
        self.value = value

def foo() -> Optional[Dummy]:
    # Return a Dummy object with a 80% chance
    if random.random() < 0.8:
        return Dummy(1)
    else:
        return None

def main() -> None:
    # (0.8)^3 = 0.512 => 51.2% chance that all three dummies are not None
    dummy_1, dummy_2, dummy_3 = foo(), foo(), foo()

    if dummy_1 is None or dummy_2 is None or dummy_3 is None:
        return

    string = "LOL"
    print(string[dummy_1.value - 1], end="")
    print(string[dummy_2.value], end="")
    print(string[dummy_3.value + 1])


if __name__ == "__main__":
    main()
```

This code has approximately a 50% chance to print "LOL" when run, nothing wrong here; the `None` checks are there to avoid the exception: `AttributeError: 'NoneType' object has no attribute 'value'`.
Without the `if dummy_1 is None or ...` check, `pyright` correctly reports an `reportOptionalMemberAccess` issue on all three objects.
But if I substitute the expression with the equivalent:
```python
if not all([dummy_1, dummy_2, dummy_3]):
    return
```
then all of the three accesses get reported:
![image](https://github.com/user-attachments/assets/e630558a-8e5d-4392-a75d-9830918c350e)

even though the code (and the null checks) is practically the same. Is this intended behavior? I know that `all()` checks for truthy values (and that `all([123, 0])` would still evaluate to `False`, even though 0 is not `None`), but in the case of class object I think it makes perfect sense.

I am using `pyright` trough the `pylance` extension, version 2024.9.1, on the latest version of VSCode. Here is more info about my install:
Version: 1.93.0 (system setup)
Commit: 4849ca9bdf9666755eb463db297b69e5385090e3
Date: 2024-09-04T13:02:38.431Z
Electron: 30.4.0
ElectronBuildId: 10073054
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Windows_NT x64 10.0.19045

Python version: 3.12.5

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The expression `all([dummy_1, dummy_2, dummy_3])` is not a supported type guard pattern. For a full list of supported type guard patterns, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

For more complex type guards, you can write your own user-defined type guard function using `TypeIs`. For details, refer to [this documentation](https://typing.readthedocs.io/en/latest/spec/narrowing.html#typeis).

**Comment by belliriccardo** [comment]
Ok, I'll check it out in more detail.
Thanks for the response and your time!
